### ISSUE TITLE
Implement function error! Introduce varieties, to make it easy to print messages

### ISSUE THREAD

**Issue by yossigil** [yossigil/safot#354]
Print dumps and other debugging information.

**Comment by yossigil** [comment]
I presume that we should consume more input in this case. @yossigil

**Comment by yossigil** [comment]
Of course, at this stage, let the error function throw an exception, and integer which is the line number, and
catch it in the main loop,.

**Comment by DorYeheskel** [comment]
In REPL there are 2 kinds of errors:
1. In the Read stage - syntax error (e.g unbalanced parenthesis)
2. In the Evaluate stage - evaluation error (e.g cons with 1 parameter)

This issue is about printing debugging information about 1 ? correct ? 

For example:
```bash
> ./mini_lisp
> ( ) ) 
line 1: syntax error             
> 
```

@yossigil

**Comment by yossigil** [comment]
The issue is about introducing varieties to the error() function. Not too many I think are needed.

**Comment by yossigil** [comment]
Ok, so the official issue is resolved; see fluenton error in the code and its idioms

**Comment by DorYeheskel** [comment]
@yossigil 
I think we still need to have full error messages + test for each of the 9 types of errors  + test for the output messages.
I'm still working on it. 

I will open another issue for these 3 missions:
1. **Full error messages as written in mini-lisp doc:**
For example, we need to identify between "car of atom" and "car of nil", 
by printing (before the "throw" in the error function) different messages to stderr, for each of these 2 cases (I'm still working on this). 
Currently in RELP it prints "CAR" for both of the errors, so in addition to print err.car and err.cdr, we need to print some message to stderr before that.

2. **test-error** which will check all the 9 types of errors in mini-lisp - using your new macro : EXPECT_EXCEPTION
For example, testing "UnknownLambda": EXPECT_EXCEPTION( S("(f 'a)").eval(),   S(""),   UNKNOWN);

3. **test-error-outputs**  - using redirection of stderr to buffer and check it with EXPECT_EQ

Please let me know what you think of this.


Notes:
I had to change my implementations few times in the last 3 days because we worked on the same files in parallel, so some of the push that you did cause me merge conflicts - CI is necessary or some sort of mechanism to avoid these conflicts in the future.

**Comment by DorYeheskel** [comment]
Plus, about the 9 types of errors in mini-lisp and how to test them:

```cpp
CarOfAtom         - S("(car 'a')").eval()
CarOfNil          - S("(car ())").eval()
CdrOfAtom         - S("(cdr t)").eval()
CdrOfNil          - S("(cdr ())").eval()
UnboundVariable   - S("(eval 'a)").eval()
MissingNames      - bind(S("('a 'b)"), S("('x 'y 'z)")
MissingValues     - bind(S("('a 'b 'c)"), S("('x 'y)")
UnknownLambda     - S("(f 'a)").eval()
OtherError        - S("(error 'a)").eval()
```

**Comment by yossigil** [comment]
OOPS, I missed this.
I implemented a small error handling stuff, which throws a pair with all the information.
I currently have tests for All but the last two. Check it out, look for EXPECT_EXCEPTION in test.h, and 
it use in other files.
